// @oslo-flanders/core/lib/utils/namespaces.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ns = void 0;
const rdf_data_factory_1 = require("rdf-data-factory");
const factory = new rdf_data_factory_1.DataFactory();
// TODO: remove example.org from prefixes
var Prefixes;
(function (Prefixes) {
    Prefixes["example"] = "http://example.org/";
    Prefixes["adms"] = "http://www.w3.org/ns/adms#";
    Prefixes["dcat"] = "http://www.w3.org/ns/dcat#";
    Prefixes["dcterms"] = "http://purl.org/dc/terms/";
    Prefixes["foaf"] = "http://xmlns.com/foaf/0.1/";
    Prefixes["owl"] = "http://www.w3.org/2002/07/owl#";
    Prefixes["vlaanderen"] = "https://data.vlaanderen.be/ns/";
    Prefixes["void"] = "http://rdfs.org/ns/void#";
    Prefixes["rdf"] = "http://www.w3.org/1999/02/22-rdf-syntax-ns#";
    Prefixes["rdfs"] = "http://www.w3.org/2000/01/rdf-schema#";
    Prefixes["qb"] = "http://purl.org/linked-data/cube#";
    Prefixes["skos"] = "http://www.w3.org/2004/02/skos/core#";
    Prefixes["xsd"] = "http://www.w3.org/2001/XMLSchema#";
    Prefixes["person"] = "http://www.w3.org/ns/person#";
    Prefixes["vann"] = "http://purl.org/vocab/vann/";
    Prefixes["shacl"] = "http://w3.org/ns/shacl#";
    Prefixes["prov"] = "http://www.w3.org/ns/prov#";
})(Prefixes || (Prefixes = {}));
const vocab = () => {
    const namespaces = {};
    for (const prefix in Prefixes) {
        const expansion = Prefixes[prefix];
        namespaces[prefix] = (localName = '') => factory.namedNode(expansion + localName);
    }
    return namespaces;
};
exports.ns = vocab();
