// n3/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "BlankNode", {
  enumerable: true,
  get: function () {
    return _N3DataFactory.BlankNode;
  }
});
Object.defineProperty(exports, "DataFactory", {
  enumerable: true,
  get: function () {
    return _N3DataFactory.default;
  }
});
Object.defineProperty(exports, "DefaultGraph", {
  enumerable: true,
  get: function () {
    return _N3DataFactory.DefaultGraph;
  }
});
Object.defineProperty(exports, "EntityIndex", {
  enumerable: true,
  get: function () {
    return _N3Store.N3EntityIndex;
  }
});
Object.defineProperty(exports, "Lexer", {
  enumerable: true,
  get: function () {
    return _N3Lexer.default;
  }
});
Object.defineProperty(exports, "Literal", {
  enumerable: true,
  get: function () {
    return _N3DataFactory.Literal;
  }
});
Object.defineProperty(exports, "NamedNode", {
  enumerable: true,
  get: function () {
    return _N3DataFactory.NamedNode;
  }
});
Object.defineProperty(exports, "Parser", {
  enumerable: true,
  get: function () {
    return _N3Parser.default;
  }
});
Object.defineProperty(exports, "Quad", {
  enumerable: true,
  get: function () {
    return _N3DataFactory.Quad;
  }
});
Object.defineProperty(exports, "Reasoner", {
  enumerable: true,
  get: function () {
    return _N3Reasoner.default;
  }
});
Object.defineProperty(exports, "Store", {
  enumerable: true,
  get: function () {
    return _N3Store.default;
  }
});
Object.defineProperty(exports, "StoreFactory", {
  enumerable: true,
  get: function () {
    return _N3StoreFactory.default;
  }
});
Object.defineProperty(exports, "StreamParser", {
  enumerable: true,
  get: function () {
    return _N3StreamParser.default;
  }
});
Object.defineProperty(exports, "StreamWriter", {
  enumerable: true,
  get: function () {
    return _N3StreamWriter.default;
  }
});
Object.defineProperty(exports, "Term", {
  enumerable: true,
  get: function () {
    return _N3DataFactory.Term;
  }
});
Object.defineProperty(exports, "Triple", {
  enumerable: true,
  get: function () {
    return _N3DataFactory.Triple;
  }
});
exports.Util = void 0;
Object.defineProperty(exports, "Variable", {
  enumerable: true,
  get: function () {
    return _N3DataFactory.Variable;
  }
});
Object.defineProperty(exports, "Writer", {
  enumerable: true,
  get: function () {
    return _N3Writer.default;
  }
});
exports.default = void 0;
Object.defineProperty(exports, "getRulesFromDataset", {
  enumerable: true,
  get: function () {
    return _N3Reasoner.getRulesFromDataset;
  }
});
Object.defineProperty(exports, "termFromId", {
  enumerable: true,
  get: function () {
    return _N3DataFactory.termFromId;
  }
});
Object.defineProperty(exports, "termToId", {
  enumerable: true,
  get: function () {
    return _N3DataFactory.termToId;
  }
});
var _N3Lexer = _interopRequireDefault(require("./N3Lexer"));
var _N3Parser = _interopRequireDefault(require("./N3Parser"));
var _N3Writer = _interopRequireDefault(require("./N3Writer"));
var _N3Store = _interopRequireWildcard(require("./N3Store"));
var _N3StoreFactory = _interopRequireDefault(require("./N3StoreFactory"));
var _N3Reasoner = _interopRequireWildcard(require("./N3Reasoner"));
var _N3StreamParser = _interopRequireDefault(require("./N3StreamParser"));
var _N3StreamWriter = _interopRequireDefault(require("./N3StreamWriter"));
var Util = _interopRequireWildcard(require("./N3Util"));
exports.Util = Util;
var _N3DataFactory = _interopRequireWildcard(require("./N3DataFactory"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
// Named exports
// Export all named exports as a default object for backward compatibility
var _default = exports.default = {
  Lexer: _N3Lexer.default,
  Parser: _N3Parser.default,
  Writer: _N3Writer.default,
  Store: _N3Store.default,
  StoreFactory: _N3StoreFactory.default,
  EntityIndex: _N3Store.N3EntityIndex,
  StreamParser: _N3StreamParser.default,
  StreamWriter: _N3StreamWriter.default,
  Util,
  Reasoner: _N3Reasoner.default,
  DataFactory: _N3DataFactory.default,
  Term: _N3DataFactory.Term,
  NamedNode: _N3DataFactory.NamedNode,
  Literal: _N3DataFactory.Literal,
  BlankNode: _N3DataFactory.BlankNode,
  Variable: _N3DataFactory.Variable,
  DefaultGraph: _N3DataFactory.DefaultGraph,
  Quad: _N3DataFactory.Quad,
  Triple: _N3DataFactory.Triple,
  termFromId: _N3DataFactory.termFromId,
  termToId: _N3DataFactory.termToId
};
