// @oslo-flanders/core/lib/logging/Logger.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseLogger = void 0;
require("reflect-metadata");
const inversify_1 = require("inversify");
let BaseLogger = class BaseLogger {
    error(message) {
        return this.log('error', message);
    }
    warn(message) {
        return this.log('warn', message);
    }
    info(message) {
        return this.log('info', message);
    }
    verbose(message) {
        return this.log('verbose', message);
    }
    debug(message) {
        return this.log('debug', message);
    }
    silly(message) {
        return this.log('silly', message);
    }
};
BaseLogger = __decorate([
    (0, inversify_1.injectable)()
], BaseLogger);
exports.BaseLogger = BaseLogger;
